File: Programmers/[17686] 파일명 정렬/ONE/17686.java
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class File {

    String head;
    int number;
    String origin;

    public File(String file) {
        this.origin = file;
        file = file.toLowerCase();

        Pattern pattern = Pattern.compile("\\d+|\\D+");
        Matcher matcher = pattern.matcher(file);

        matcher.find();
        this.head = matcher.group(0);

        matcher.find();
        this.number = Integer.parseInt(matcher.group(0));
    }
}

class Solution {
    public String[] solution(String[] files) {
        ArrayList<File> list = new ArrayList<>();
        ArrayList<String> answer = new ArrayList<>();

        for (String file : files)
            list.add(new File(file));

        list.sort((o1, o2) -> {
            if (o1.head.compareTo(o2.head) == 0)
                return o1.number - o2.number;
            return o1.head.compareTo(o2.head);
        });

        for(File file : list)
            answer.add(file.origin);

        return answer.toArray(new String[0]);
    }
}

File: Programmers/[17686] 파일명 정렬/KingDonggyu/17686.js
function solution(files) {
  const fileNames = files.map((file) => {
    const name = ['', '', '']; // head, number, tail
    let isAfterNumber = false;

    for (const f of file) {
      if (!name[2] && !isNaN(parseInt(f))) {
        name[1] += f;
        isAfterNumber = true;
        continue;
      }

      if (isAfterNumber) {
        name[2] += f;
        continue;
      }

      name[0] += f;
    }

    return name;
  });

  fileNames.sort((a, b) => {
    const headA = a[0].toLowerCase();
    const headB = b[0].toLowerCase();

    if (headA === headB) {
      return +a[1] - +b[1];
    }

    return headA > headB ? 1 : -1;
  });

  return fileNames.map((fileName) => fileName.join(''));
}